src/main/java/de/tum/in/www1/exerciseapp/service/CourseService.java

```diff
@@ -105,4 +103,80 @@ public void delete(Long id) {
+     *
+     * @param courseId the courseId
+     * @return the collection of results in the result score the average score is saved, which contains the participation and the user
+     */
+    @Transactional(readOnly = true)
+    public Collection<Result> getAllOverallScoresOfCourse(Long courseId) {
+        Course course = findOne(courseId);
+        Set<Exercise> exercisesOfCourse = course.getExercises();
+        //key stores the userId to identify if he already got a score, value contains the Result itself with the score of the user
+        HashMap<Long, Result> allOverallSummedScoresOfCourse = new HashMap<>();
+
+        for (Exercise exercise : exercisesOfCourse) {
+            Set<Participation> participations = exercise.getParticipations();
+            for (Participation participation : participations) {
+
+                //id of user in the datbase to reference to the user
+                long userID = participation.getStudent().getId();
+                Result bestResult = bestResultScoreInParticipation(participation);
+
+                //if student already appeared, once add the new score to the old one
+                if (allOverallSummedScoresOfCourse.containsKey(userID)) {
+                    long oldScore = allOverallSummedScoresOfCourse.get(userID).getScore();
+                    bestResult.setScore(oldScore + bestResult.getScore());
+                    allOverallSummedScoresOfCourse.remove(userID);
+                }
+                allOverallSummedScoresOfCourse.put(userID, bestResult);
+            }
+        }
+
+        //divide the scores by the amount of exercises to get the average Score of all Exercises
+        Collection<Result> allOverallScores = allOverallSummedScoresOfCourse.values();
+        int numberOfExercises = exercisesOfCourse.size();
+        for (Result result : allOverallScores) {
+            result.setScore(result.getScore() / (long) numberOfExercises);
+        }
+
+        return allOverallScores;
+    }
+
+    /**
+     * Find the best Result in a Participation
+     *
+     * @param participation the participation you want the best result from
+     * @return the best result a student had within the time of the exercise
+     */
+    @Transactional(readOnly = true)
+    public Result bestResultScoreInParticipation(Participation participation) {
+        Set<Result> results = participation.getResults();
+
+        Result bestResult = null;
+        for (Result result : results) {
+            if (bestResult == null) {
+                bestResult = new Result();
+                bestResult.setScore(result.getScore());
+            } else if (bestResult.getScore() != null
+                && result.getScore() != null //score can be stored as null in the DB
+                && bestResult.getScore() < result.getScore()
+                && participation.getExercise().getDueDate().isBefore(result.getCompletionDate())//only use results which are completed before the dueDate
```

Comment:
I think this else if statement is not 100% correct, because in case a result gets worse within the due date, it would not be used. Please do it differently: 
**Case 1: there is a due date:**
first determine the latest result within the due date (i.e. the one that is closest to the due date, but still before it), and simply use this result
**Case 2: there is no due date:**
simply take the result with the latest completion date